Service/src/main/java/com/epam/esm/service/impl/GiftCertificateServiceImpl.java

```diff
@@ -1,42 +1,58 @@
 package com.epam.esm.service.impl;
 
 import com.epam.esm.dao.GiftCertificateDao;
-import com.epam.esm.dao.TagDao;
 import com.epam.esm.entity.GiftCertificate;
 import com.epam.esm.entity.Tag;
+import com.epam.esm.exception.EntityException;
 import com.epam.esm.service.GiftCertificateService;
+import com.epam.esm.service.TagService;
 import com.epam.esm.validator.GiftCertificateValidator;
+import lombok.extern.log4j.Log4j2;
 import org.springframework.beans.factory.annotation.Autowired;
 import org.springframework.stereotype.Service;
 import org.springframework.transaction.annotation.Transactional;
 
 import java.util.HashMap;
 import java.util.List;
 import java.util.Map;
+import java.util.Set;
 
+import static com.epam.esm.exception.ExceptionCode.*;
+
+@Log4j2
 @Service
 public class GiftCertificateServiceImpl implements GiftCertificateService {
     private final GiftCertificateDao giftCertificateDao;
-    private final TagDao tagDao;
+    private final TagService tagService;
     private final GiftCertificateValidator giftCertificateValidator;
-    private final static String FIELD_NAME = "name";
-    private final static String FIELD_DESCRIPTION = "description";
-    private final static String FIELD_PRICE = "price";
-    private final static String FIELD_DURATION = "duration";
+    private static final String FIELD_NAME = "name";
+    private static final String FIELD_DESCRIPTION = "description";
+    private static final String FIELD_PRICE = "price";
+    private static final String FIELD_DURATION = "duration";
 
     @Autowired
-    public GiftCertificateServiceImpl(GiftCertificateDao giftCertificateDao, TagDao tagDao,
+    public GiftCertificateServiceImpl(GiftCertificateDao giftCertificateDao, TagService tagService,
                                       GiftCertificateValidator giftCertificateValidator) {
         this.giftCertificateDao = giftCertificateDao;
-        this.tagDao = tagDao;
         this.giftCertificateValidator = giftCertificateValidator;
+        this.tagService = tagService;
     }
 
     @Override
-    public void create(GiftCertificate giftCertificate) {
-        if (giftCertificateValidator.isValid(giftCertificate)){
-        giftCertificateDao.create(giftCertificate);
+    @Transactional
+    public GiftCertificate create(GiftCertificate giftCertificate) {
+        if (!giftCertificateValidator.isValid(giftCertificate)){
+            throw new EntityException(NOT_VALID_GIFT_CERTIFICATE_DATA.getErrorCode());
+        }
+        Long newId = giftCertificateDao.create(giftCertificate);
+        Set<Tag> tags = giftCertificate.getTags();
+        if(tags != null && tags.size() != 0){
```

Comment:
please update validation service to validate list of tags

Reply:
Done